src/lsptoolshost/buildDiagnosticsService.ts

```diff
@@ -0,0 +1,146 @@
+        // reported by the build to show build-only problems.
+        if (currentFileBuildDiagnostics) {
+            const buildDiagnostics = BuildDiagnosticsService.filerDiagnosticsFromBuild(
+                currentFileBuildDiagnostics[1],
+                buildOnlyIds,
+                true
+            );
+            this._diagnosticsReportedByBuild.set(uri, buildDiagnostics);
+        }
+    }
+
+    public static filerDiagnosticsFromBuild(
+        diagnosticList: vscode.Diagnostic[],
+        buildOnlyIds: string[],
+        isDocumentOpen: boolean
+    ): vscode.Diagnostic[] {
+        const configuration = vscode.workspace.getConfiguration();
+        const analyzerDiagnosticScope = configuration.get(
+            'dotnet.backgroundAnalysis.analyzerDiagnosticsScope'
+        ) as AnalysisSetting;
+        const compilerDiagnosticScope = configuration.get(
+            'dotnet.backgroundAnalysis.compilerDiagnosticsScope'
+        ) as AnalysisSetting;
+
+        // If compiler and analyzer diagnostics are set to "none", show everything reported by the build
+        if (analyzerDiagnosticScope === AnalysisSetting.None && compilerDiagnosticScope === AnalysisSetting.None) {
+            return diagnosticList;
+        }
+
+        // Filter the diagnostics reported by the build. Some may already be shown by live diagnostics.
+        const buildOnlyDiagnostics: vscode.Diagnostic[] = [];
+        diagnosticList.forEach((d) => {
+            if (d.code) {
+                // If it is a project system diagnostic (e.g. "Target framework out of support")
+                // then always show it. It cannot be reported by live.
+                if (this.isProjectSystemDiagnostic(d)) {
+                    buildOnlyDiagnostics.push(d);
+                }
+                // If it is a "build-only"diagnostics (i.e. it can only be found by building)
+                // then always show it. It cannot be reported by live.
+                else if (buildOnlyIds.find((b_id) => b_id === d.code)) {
+                    buildOnlyDiagnostics.push(d);
+                } else {
+                    const isAnalyzerDiagnostic = BuildDiagnosticsService.isAnalyzerDiagnostic(d);
+                    const isCompilerDiagnostic = BuildDiagnosticsService.isCompilerDiagnostic(d);
+
+                    if (
+                        (isAnalyzerDiagnostic && analyzerDiagnosticScope === AnalysisSetting.None) ||
+                        (isCompilerDiagnostic && compilerDiagnosticScope === AnalysisSetting.None)
+                    ) {
+                        // If live diagnostics are completely turned off for this type, then show the build diagnostic
+                        buildOnlyDiagnostics.push(d);
+                    } else if (isDocumentOpen) {
+                        // no-op. The document is open and live diagnostis are on. This diagnostic is already being shown.
+                    } else if (
```

Comment:
This flow is a bit hard to follow.  Wondering if we could make it a bit cleaner by doing an upfront grouping of the diagnostics collection into project system diags, build only diags, compiler diags, analyzer diags.

Then iterate each group separately with hopefully simpler logic blocks
e.g.
```
const projectSystemDiagnostics = ...
const buildOnlyDiagnostics = ...
const compilerDiagnostics = ...
const analyzerDiagnostics = ...

buildOnlyDiagnostics.push(projectSystemDiagnostics);
buildOnlyDiagnostics.push(buildOnlyDiagnostics);

if ( compiler scope is none or (scope = open doc and doc not open) {
    buildOnlyDiagnostics.push(compilerDiagnostics);
}

// same for analysis
```

Reply:
I think you make a great point! I structured it this way so we don't loop through the diagnostics so many times, but that probably isn't a concern. If a project has millions of diagnostics, possible slowness in the problems window is probably the least of the user's problems.